### DISCUSSION TITLE
🚀 [v0.3.30] Milestone Release: Multimodal Pipeline Refactor & Hybrid/Recurrent Model Architecture Support

### DISCUSSION THREAD

**Discussion by JamePeng** [JamePeng/llama-cpp-python#73] [Announcements]
# 🚀 [v0.3.30] Milestone Release

## Multimodal Pipeline Refactor & Hybrid Model Architecture Support (v0.3.28 - v0.3.30)

Changelog see: https://github.com/JamePeng/llama-cpp-python/blob/main/CHANGELOG.md#0330-milestone-release

I am excited to announce the release of **llama-cpp-python v0.3.30**.

Looking back at the sampler refactoring logic in versions 0.3.24-0.3.25, and the performance and memory management improvements in versions 0.3.26-0.3.27, the long-standing issues with releasing underlying memory and VRAM were resolved. Deep function-level optimizations of CPU processing time (reducing unnecessary or repetitive Python processing) were implemented, which is extremely helpful for memory-constrained Mac products.

The reason for skipping several versions from 0.3.27 to release 0.3.30 is that these improvements were implemented sequentially, are interdependent, and their merging results in higher runtime efficiency.

All of this is to align with the official underlying runtime logic of llama.cpp and achieve better performance and runtime compatibility.

Over the past few updates (from `0.3.28` to `0.3.30`), I have significantly restructured the core engine to support the next generation of **Hybrid/Recurrent Models (e.g., Qwen3-Next(Qwen3-Coder-Next), Qwen3.5, LFM2-VL)** and introduced a highly optimized **Concurrent Multimodal Pipeline**.

If your applications involve complex multimodal contexts, hybrid architectures, or long multi-turn conversations, this update provides essential structural improvements for stability and performance.

Here is a breakdown of the key engineering updates across these releases:

---

## ⚙️ 1. Concurrent Multimodal Processing Pipeline & `MTMDChatHandler`

Evaluating prompts with multiple images previously caused CPU decoding bottlenecks. I have extracted the multimodal logic into a generic `MTMDChatHandler` base class and restructured the pipeline to process media much more efficiently:

* **Concurrent Media Decoding:** Implemented a thread-safe `ThreadPoolExecutor` to parallelize image/audio decoding. This allows the C++ backend to utilize multi-core CPUs for preprocessing, significantly reducing I/O latency for multi-image prompts.
    * *Thread count calculation method:* `max_workers = min(llama.n_threads, len(image_urls))`

* **Chronological Alignment:** By utilizing pre-allocated array indexing, I guarantee that `bitmaps` remain perfectly aligned with the user's original input order, despite asynchronous decoding. This prepares the pipeline for future video-frame processing.
* **Simplified Subclassing:** Subclasses (e.g., MiniCPM, GLM-V, Qwen2.5-VL, Qwen3-VL, and the newly merged **Qwen3.5**) now only need to define their prompt templates. I also implemented strict `**kwargs` validation and dynamic logging across all subclasses to improve developer experience.

## 🧠 2. "Negative Reverse Vocabulary" & O(1) Prefix Matching

To handle multi-turn multimodal conversations efficiently, I have improved how the engine matches current images against the KV cache history.

* **The Mechanism:** Instead of using slow dictionary lookups or identical placeholders that cause cache poisoning, we introduced a deterministic Hash-to-Negative-Integer mapping. When a media chunk is parsed, it generates a unique negative `media_id` (e.g., `-485712`) within a reserved physical space (`-100` to `-16,777,316`).
* **The Result:** This strictly isolates media tokens from the LLM's positive vocabulary space, preventing matrix collisions. It allows Python to execute native, ultra-fast **O(1) integer array comparisons** for prefix matching via `longest_token_prefix`. I also implemented logic to physically erase trailing "ghost" tokens from the C++ KV cache to prevent attention misalignment.

## 🧩 3. Hybrid/Recurrent Model State Management

Models with Hybrid/Recurrent architectures (like Qwen3.5) cannot simply truncate their KV cache due to hidden RNN states. I implemented a dedicated state management system for them starting in `0.3.28`.

* **`HybridCheckpointCache`:** A stateless sequence manager that tracks RNN state snapshots, using cryptographic SHA-256 prefix hashing to guarantee state integrity upon rollback.
* **End-of-Turn Checkpointing:** Instead of saving the massive RNN state aggressively during mid-turn media evaluation, I implemented "End-of-Turn" anchoring. A complex multi-image prompt now consumes exactly 1 LRU slot.
* **Adaptive Periodic Checkpointing:** During large text pre-filling, `eval` will dynamically scale the save frequency (max 3 triggers per eval) to minimize I/O stuttering. I also added strict state size validation before restoration to prevent buffer overflows.

## 🛡️ 4. Core `generate` & `eval` Overhaul and OOM Defense

We fundamentally overhauled the `generate` and `eval` functions to integrate hybrid model support and enhance safety mechanisms when the context window reaches its physical limits.

* **Generate Loop Integration:** Integrated `HybridCheckpointCache` directly into the `generate` loop to support seamless state rollback for recurrent architectures. Wrapped the generation loop in a `try...finally` block to guarantee safe checkpoint saving.
* **Eval Vectorization:** Adapted `eval` to use the newly vectorized `LlamaBatch.add_sequence` API (replacing the old `set_batch`), enabling granular `add_token` capabilities and dynamic `logits_array` configuration for better multi-sequence and logit control.
* **Logits Refresh Bug Fix:** Fixed the full prefix match bug in `generate` by forcing a 1-token re-evaluation to properly refresh logits.
* **Speculative Decoding Safety:** Explicitly disabled speculative decoding (draft models) for hybrid architectures to prevent irreversible state pollution in the RNN hidden layers.
* **Physical Context Shift:** The dispatch engine now preemptively calculates token boundaries. If it detects an impending OOM, it executes a mathematically safe Context Shift in `eval`—discarding the oldest unpinned tokens from both the physical C++ KV cache and the Python virtual ledger simultaneously. (Note: Hybrid architectures are not supported yet.).
* **Architecture-Safe Pre-flight Checks:** Added `memory_can_shift()` checks to proactively intercept operations on architectures that physically forbid shifting (like certain M-RoPE or `n_pos_per_embd > 1` setups), gracefully aborting to prevent fatal `GGML_ASSERT` crashes.
* **Recoverable KV Exhaustion:** Updated the `llama_decode` wrapper to treat Code 1 (KV slot exhaustion) as a recoverable return value. The engine now gracefully handles VRAM fragmentation via dynamic batch halving, with explicit guards to prevent halving below a batch size of 1.

---

### 🎉 Acknowledgments & Community Contributions

A massive thank you to the community members who made this architectural overhaul possible:

* **@alcoftTAO**: For the rapid implementation of the `Qwen35ChatHandler`, confirming our Qwen3.5 compatibility right at launch!
* **@yamikumo-DSD**: For providing rigorous code reviews, detailed testing, and complete log feedback that was crucial for debugging the pipeline.
* **@roj234**: For identifying an uncaught `ggml` check exception, which directly led to our improved `memory_can_shift()` pre-flight safety mechanism.

**Upgrade now:**

```bash
pip install "llama-cpp-python @ git+https://github.com/JamePeng/llama-cpp-python.git"
```

**Github Action Build: [Release](https://github.com/JamePeng/llama-cpp-python/releases)**

Best regards<br>
**JamePeng**

**Comment by yamikumo-DSD** [comment]
It works!
<img width="872" height="1549" alt="スクリーンショット 2026-03-03 19 23 39" src="https://github.com/user-attachments/assets/b3cb9ea9-a097-41f9-8ad3-4550c0113c4d" />
